Sources/CoreMetrics/Metrics.swift

```diff
@@ -319,19 +323,46 @@ public protocol MetricsFactory {
     ///     - label: The label for the CounterHandler.
     ///     - dimensions: The dimensions for the CounterHandler.
     func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler
+
     /// Create a backing `RecorderHandler`.
     ///
     /// - parameters:
     ///     - label: The label for the RecorderHandler.
     ///     - dimensions: The dimensions for the RecorderHandler.
     ///     - aggregate: Is data aggregation expected.
     func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler
+
     /// Create a backing `TimerHandler`.
     ///
     /// - parameters:
     ///     - label: The label for the TimerHandler.
     ///     - dimensions: The dimensions for the TimerHandler.
     func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler
+
+    /// Signals the `MetricsFactory` that the passed in `MetricHandler` will no longer be updated.
+    /// Implementing this functionality is _optional_, and depends on the semantics of the concrete `MetricsFactory`.
+    ///
+    /// In response to this call, the factory _may_ release resources associated with this metric,
+    /// e.g. in case the metric contains references to "heavy" resources, such as file handles, connections,
+    /// or large in-memory data structures.
+    ///
+    /// # Intended usage
+    ///
+    /// Metrics library implementations are _not_ required to act on this signal immediately (or at all).
+    /// However, the presence of this API allows middle-ware libraries wanting to emit metrics for resources with
+    /// well-defined life-cycles to behave pro-actively, and signal when a given metric is known to not be used anymore,
+    /// which can make an positive impact with regards to resource utilization in case of metrics libraries which keep
+    /// references to "heavy" resources.
+    ///
+    /// It is expected that some metrics libraries, may choose to omit implementing this functionality.
+    /// One such example may be a library which directly emits recorded values to some underlying shared storage engine,
+    /// which means that the `MetricHandler` objects themselves are light-weight by nature, and thus no lifecycle
+    /// management and releasing of such metrics handlers is necessary.
+    func remove<M: Metric>(metric: M)
+    // alternatively, delete the Metric type and provide overloads:
+    // func remove<C: Counter>(_ counter: C)
+    // func remove<R: Recorder>(_ recorder: R)
+    // func remove<T: Timer>(_ timer: T)
```

Comment:
i slightly prefer to be symmetric with the `makeXXX(label:)` APIs and offer `destroyXXX(label)` instead of the generic version. this could also allow us to get rid of the `Metric` protocol to keep things simple(r). wdyt?

Reply:
@ktoso after sleeping on this some i now even more in favor of the above suggestion for the following reasons:
1. api symmetry 
2. no need in marker protocol 
3. less coupling and confusion between the handler type that the backend libraries deal with (eg CounterHandler) and the metrics object which is what the apps and libraries deal with (eg  Counter). iow, the backend libraries should only deal with handler types, and as such it means a release function that takes a metric object will need to poke a hole  in the abstraction which is not very nice 
4. it opens the door to releasing by an expression which could be useful for middleware type libraries that have a bunch of metrics per connection or request and a naming convention around that

Reply:
I'm also in favour of anything that doesn't use the weird marker protocol. `makeXXX` and `destroyXXX` sound good to me!

Reply:
or obviously `myMetric.destroy()`, possibly I like this even more...

Reply:
Hmm, some good points, though I'm worried about the `destroyCounter(label: "bla")` losing type-safety and forces more of the "get it right" dance onto users / libraries where as a `destroyCounter(counter)` helps to get it right by compilation; 

Example:

- imagine if you refactor a metrics from counter to gauge `let myThingy = Gauge(...) // was Counter` -- there you had to since you want to get the diff type
- you could forget to change the `destroyCounter(myThingy.label)` call, even if you changed the make site

With accepting `destroyCounter(myThingy)` you would get a compile error and would notice that you'd almost messed up, but the compiler helped you spot that.

Reply:
Hah, I now also realized that the same was true for the `release(Metric)`, so perhaps if we like the 3 functions for the right types that would be a win-win situation?

Reply:
@ktoso you make a great point regarding type safety. however point 3 remains an challenge for me. imo, the application/middleware that will be doing the "releasing" (call site) should only ever know about the metric objects (eg `Counter`, `Timer`) and the backend library should only ever know about the handlers (eg `CounterHandler`, `TimerHandler`) and the proposal atm breaks this separation of concerns. the one common things between them is the label, which is why i arrived at my alternative proposal of releasing by label

@weissi proposal can work too, because the metric object (eg `Counter`, `Timer`) has direct access to the underlying handler (eg `CounterHandler`, `TimerHandler`) by definition so the separation of concerns remains intact

Reply:
/cc @hartbit which had good feedback on logging API, maybe can help here too

Reply:
Hmm, I see what you mean -- that if I don't release I (user) only touch Counter, but if I want to release I suddenly need to reach into all the way to Factory. 

Let me see how it could look like. Main concern was the requirement of additional storage for the reference to the factory, but perhaps -- and also since we enforce that "bootstrap only once" deletes could reach through the shared field...? A bit iffy... could cause weird issues of "release in wrong logging library instance" happening 🤔 

I'll give it a shot as separate commit and we can look at it here

Reply:
Is there a reason we don't benefit from the fact that all `...Handler` protocols are reference types (`AnyObject`) to ask backend libraries to do their cleanup in `deinit`?

Reply:
Thanks for chiming in @hartbit :)
Sadly the `deinit` style has a lot of insane corner cases, and is somewhat abusing the mechanism (the lifecycles of "the metric" do not necessarily match 1:1 "the object"'s lifecycle), I explored it in the writeup over here: https://github.com/apple/swift-metrics/issues/2

I think we are getting to a final design here slowly... we'll be able to:

- keep full typesafety of the releases -- no need to cast
- no crossing boundaries between handler+factory (impl) and counter+system (user api)
- destroy() _can_ live on the objects themselves, leading to simple use and discovery of the feature
- only place to care about "implementing destroying" remains factory, which semantically makes sense

Reply:
Thanks for the explanation! It's true that ARC does come with design challenges, but I'm also worried about a `destroy` function:

* Because destroy will sometimes be a no-op, sometimes not (depending on the implementation), users might misuse them but think everything is fine (because they are using a backend that is a no-op) but start having issues as soon as they switch backends.
* A destroy function doesn't protect users from continuing to use the metrics objects after being destroyed.

While using ARC might mean abusing it slightly, and there cause slightly more headaches in the implementation aspect of things, I think it comes with greater simplicity for the user.

Reply:
/cc @weissi who was part of the previous ARC discussion